inst/chunks/fun_xbetax.stan

```diff
@@ -0,0 +1,118 @@
+  /* Extended support beta log-PDF of a single response
+   * Args:
+   *   y: the response value
+   *   mu: mean parameter of the beta distribution
+   *   phi: precision parameter of the beta distribution
+   *   kappa: the exceedance parameter (u in Kosmidis & Zeileis, 2024, https://arxiv.org/abs/2409.07233)
+   * Returns:
+   *   a scalar to be added to the log posterior
+   */
+
+   real xbetax_lpdf(real y, real mu, real phi, real kappa) {
+     real denom = (1 + 2 * kappa);
+     if (y <= 0)
+       return beta_proportion_lcdf(kappa / denom | mu, phi);
+     else if (y >= 1)
+       return beta_proportion_lccdf((1 + kappa) / denom | mu, phi);
+     else
+       return beta_proportion_lpdf((y + kappa) / denom | mu, phi) - log(denom);
+   }
+
+   real xbetax_lpdf(vector y, vector mu, vector phi, vector kappa) {
+     int N = size(y);
+     int N_zer = 0;
+     int N_one = 0;
+     int N_oth = 0;
+
+     int i_zer = 0;
+     int i_one = 0;
+     int i_oth = 0;
+     
+     for (i in 1:N) {
+     	 N_zer += y[i] <= 0;
+	 N_one += y[i] >= 1;
+	 N_oth += (y[i] > 0) && (y[i] < 1);
+     }     
+
+     array[N_zer] int zer;
+     array[N_one] int one;
+     array[N_oth] int oth;
+
+     for (i in 1:N) {
+       if (y[i] <= 0) {
+         i_zer += 1;
+         zer[i_zer] = i;
+       }
+       else if (y[i] >= 1) {
+         i_one += 1;
+	 one[i_one] = i;
+       }
+       else {
+	 i_oth += 1;
+	 oth[i_oth] = i;
+       }
+     }
+
+     real ll_zer = beta_proportion_lcdf(kappa[zer] ./ (1 + 2 * kappa[zer]) | mu[zer], phi[zer]);
+     real ll_one = beta_proportion_lccdf((1 + kappa[one]) ./ (1 + 2 * kappa[one]) | mu[one], phi[one]);
+     real ll_oth = beta_proportion_lpdf((y[oth] + kappa[oth]) ./ (1 + 2 * kappa[oth]) | mu[oth], phi[oth]) - sum(log(1 + 2 * kappa[oth]));
+     real ll = ll_zer + ll_one + ll_oth;
+     return ll;	
+   }
+
+   real xbetax_lpdf(vector y, vector mu, real phi, vector kappa) {
+      vector[1] phiv;
```

Comment:
I don't think this will work as you expect. You also need to handle the fact that phi for example is not of length 1 not of length N. I believe your fully vectorized version expects all vectors to be of the same length as y. This issue holds for all these partial vector versions.

Reply:
OK. I see now. The vectorized versions in commit [e8b3352](https://github.com/paul-buerkner/brms/pull/1698/commits/e8b3352389c3949eb46e9c7f166b0157d1247080) should now work as expected. I tested things around with the `LossAversion` data (the example I had in `?brm` and will appear in the vignette) and despite it being relatively small we can already see speedups.

Here is a comparison in elapsed times between vectorized and non-vectorized implementations
<img width="1148" alt="Screenshot 2024-11-01 at 12 12 32" src="https://github.com/user-attachments/assets/8f159bd2-fdaa-45ad-845d-14c726944e4f">

where the x axis is what components are passed to `brm`, based on 20 replications per model with `init = 0`, `chains = 1`, `iter = 2000`.

The fits between vectorized and non-vectorized implementations are the same.

I can also share the script that reproduces the figure and carries out the tests on the `brmsfit` objects.  If you want that what is the right place to push it (`tests/local`?)?.

Reply:
One thing I wanted to ask: is there any value in having vectorized functions with real `mu`, e.g.

```
xbetax_lpdf(vector y, real mu, vector phi, real kappa)
```
? These are not difficult to setup, but I am not sure if `rbm` would ever pass real `mu` to the compiled STAN code

Reply:
Thank you! I will now do some edits to the PR and let you know for a final check before merging.

About your question: mu will indeed always be a vector in brms so no need to have a vectorized real mu implementation.